### ISSUE TITLE
AudioThreadPool race condition on Windows.

### ISSUE THREAD

**Issue by eldaranne** [cocos2d/cocos2d-x#14425]
_Version_
- v3.8

_Description_
- On Win32, the threads from the AudioThreadPool are created detached, this causes a possible race condition where the AudioEngineThreadPool can be destroyed and threadFunc() tries to lock _queueMutex. Since the object has been destroyed, threadFunc then crashes because it tries to access a non-existent object.

_Question_
Is there a reason to create them detached on Win32? Unless it is absolutely necessary, they should join so the race condition disappears.

**Comment by fuyifan** [comment]
@WenhaiLin Please give us a hand, we really need your help.

**Comment by fuyifan** [comment]
@pandamicro 哥哥求这个版本把这个修了吧

**Comment by fuyifan** [comment]
I found the clue to re-produce  this problem. Just try to play a big enough sound file and invoke cocos2d::experimental::AudioEngine::end(). However, if I invoke stopAll() first followed by end(), everything seems to be OK. @pandamicro @WenhaiLin

**Comment by dumganhar** [comment]
@eldaranne @fuyifan , thanks for your feedback.
I'm gonna check this problem.

**Comment by dumganhar** [comment]
> Is there a reason to create them detached on Win32? Unless it is absolutely necessary, they should join so the race condition disappears.

I also don't know it we need to `detach` just for windows platform. :(

**Comment by ghost** [comment]
Just not detaching on Win32 solved the issue for me:

```
#if CC_TARGET_PLATFORM == CC_PLATFORM_WIN32
    if (_audioEngineImpl && s_threadPool == nullptr)
    {
        s_threadPool = new (std::nothrow) AudioEngineThreadPool(false); //AudioEngineThreadPool(true); 
    }
```

I think detaching is important for brojects in which you have lots of sounds playing at the same time in order to have better performance ?

PS: in my project, it crashes while I am using very short files, so this is not linked to the sounds duration I think

**Comment by dumganhar** [comment]
Yep, it's relative to audio objects released before audio thread exits. audio thread may use a freed audio objects(AudioPlayer or AudioCache).

**Comment by dumganhar** [comment]
It was fixed in v3.14. I closed this issue.
